Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -4106,23 +4107,125 @@ std::pair<B3::PatchpointValue*, PatchpointExceptionHandle> AirIRGenerator::emitC
     return { patchpoint, exceptionHandle };
 }
 
-auto AirIRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signature, Vector<ExpressionType>& args, ResultList& results) -> PartialResult
+AirIRGenerator::CallPatchpointData AirIRGenerator::emitTailCallPatchpoint(BasicBlock* block, const Checked<int32_t>& newStackOffsetFromFP, Vector<ArgumentLocation>& locations, Vector<TypedTmp>& args, Vector<ConstrainedTmp> patchArgs)
+
 {
-    ASSERT(signature.as<FunctionSignature>()->argumentCount() == args.size());
+    auto shuffleStackArg = [this, block, newStackOffsetFromFP] (Tmp tmp, int32_t offsetFromSP) -> void {
+        Checked<int32_t> offsetFromFP = newStackOffsetFromFP + offsetFromSP;
 
-    m_makesCalls = true;
+        if (offsetFromFP < 0) {
+            StackSlot* stackSlot = m_code.addStackSlot(sizeof(Register), StackSlotKind::Locked);
+            stackSlot->setOffsetFromFP(offsetFromFP);
+            append(block, tmp.isGP() ? Move : MoveDouble, tmp, Arg::stack(stackSlot));
+            return;
+        }
 
-    for (unsigned i = 0; i < signature.as<FunctionSignature>()->returnCount(); ++i)
-        results.append(tmpForType(signature.as<FunctionSignature>()->returnType(i)));
+        append(block, tmp.isGP() ? Move : MoveDouble, tmp, Arg::addr(Tmp(GPRInfo::callFrameRegister), offsetFromFP));
+    };
+
+    auto tmp = g64();
+
+    append(block, Move, Arg::addr(Tmp(MacroAssembler::framePointerRegister), CallFrame::returnPCOffset()), tmp);
+    shuffleStackArg(tmp, -CallFrame::returnPCOffset());
```

Comment:
From CallFrame.h, we have the following comment. I would love to have a similar one here explaining how the layout changes, and perhaps a link to your blog post.

```
Layout of stack right before tail call F -> G

|          ......            |   |                                                                  |          ......            |
+----------------------------+   |                                                                  +----------------------------+
|           F.argN           |   v  lower address                      |-------------------->       |           G.argM           |
+----------------------------+                                         |                            +----------------------------+
|           arg1             |                                         |                            |           arg2             |
+----------------------------+                                         |                            +----------------------------+
|           arg0             |                                         |                            |           arg1             |
+----------------------------+                                         |                            +----------------------------+
|           F.this           |                                         |                            |           arg0             |
+----------------------------+                                         |                            +----------------------------+
| argumentCountIncludingThis |                                         |                            |           this'            |
+----------------------------+                                         |                            +----------------------------+
|  F.callee (aka F, unused in wasm) |                                  |                            |          A.C.I.T.'         |
+----------------------------+                                         |                            +----------------------------+
|        F.codeBlock         |                                         |                            |        G.callee            |
+----------------------------+                                         |                            +----------------------------+
| return-address after F     |                                         |                            |        G.codeBlock         |
+----------------------------+                                         |                            +----------------------------+
|       F.caller.FP          |                                         |                            |   return-address after F   |
+----------------------------+  <- F.FP                                |                            +----------------------------+
|          local0            |                                         |                            |          F.caller.FP       |
+----------------------------+            argM to G  ------------------|                     G.FP-> +----------------------------+
|          local1            |            ....                                                      |          G.local0          |
+----------------------------+            arg0 to G                                                 +----------------------------+
|          localN            |   <----|                                                             |          G.local1          |
+----------------------------+ <- SP  |                                                             +----------------------------+
|          ......            |        |                                                     G.SP->  |          G.localM          |
                                      |- New tmp stack slots are eventually allocated here
```

@Constellation please confirm this is correct.

Based on that diagram, why isn't this -sizeof(CallFrame)+returnPCOffset()?

Reply:
So semantically, that is still wrong, right? returnPCOffset is relative to the start of CallFrame